### PR TITLE
Implemented Ball Tree using Haversine distance as mentioned in #237

### LINKED ISSUE #237: Feature Request: BallTree algorithm with haversine distance
This is a neat polars plug in. Thanks! Do you have plans for implementing the Ball Tree algorithm like in scikit-learn? 

https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html

The haversine distance metric to find nearest neighbors between latitude/longitude points on Earth would be very useful.

### PR DESCRIPTION
**Pull request by AtlasPilotPuppy**
Resolves #237
This PR implements Ball Tree with Haversine and Euclidian distance.

Thie implmentation does not utilize the SpatialQueries Trait. But can be modified after some discussion on how to handle the differeing parameters
This does implement all the function utilized in the query_knn.py There is demo code at the bottom of basics.ipynb

There are also some Rust tests to test the functioning of the ball tree. This ball tree was [influenced by this repo](https://github.com/grantslatton/ball-tree).

The current approach was chosen since it creates a ball tree and remain in memory for iterative querying. This is very efficient in our case since these are applied to all rows of the dataframe.
This is my first time writing a Polars extension so ideas and feedback is welcome.

### CODE REVIEW

**Comment by abstractqqq** [comment]
@AtlasPilotPuppy  I am merging it, but I will make some changes. e.g. move the file for better organization, reuse some exsisting code, etc.. Thanks a lot for the work.

**Comment by AtlasPilotPuppy** [comment]
@abstractqqq much appreciated. Let me know if there are other issues you would like to look at.
I wanted to reuse more but my lack of familiarity was a hindrance.
Another thing is that the Leaf doesn't own the points array. Which leads to some interesting challenges. It may be something else to consider to have the leaf take the ownership of the points rather than the containing function.

**Comment by abstractqqq** [comment]
> @abstractqqq much appreciated. Let me know if there are other issues you would like to look at. I wanted to reuse more but my lack of familiarity was a hindrance. Another thing is that the Leaf doesn't own the points array. Which leads to some interesting challenges. It may be something else to consider to have the leaf take the ownership of the points rather than the containing function.

There is a OwnedLeaf type FYI. For KNN style query in dataframe, I use the regular Leaf type because I don't need to copy data any more. For the py_kdt model I expose to Python (in src/pymodels.rs), I use the OwnedLeaf type because when data is passed from Python to Rust, life time info is obscure. I think in general the lifetime system forces you to in-line some functions which you would normally factor out. This forces one to make only "really good factors" and in-line those that are only for convenience.. It has its pros and cons. 

Thank you again and this is great!
